pkg/querier/http.go

```diff
@@ -126,3 +134,54 @@ func (q *Querier) LabelHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+// TailHandler is a http.HandlerFunc for handling tail queries.
+func (q *Querier) TailHandler(w http.ResponseWriter, r *http.Request) {
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+
+	queryRequestPtr, err := httpRequestToQueryRequest(r)
+	if err != nil {
+		server.WriteError(w, err)
+		return
+	}
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		level.Error(util.Logger).Log("Error in upgrading websocket", fmt.Sprintf("%v", err))
+		return
+	}
+
+	defer func() {
+		err := conn.Close()
+		level.Error(util.Logger).Log("Error closing websocket", fmt.Sprintf("%v", err))
+	}()
+
+	// response from httpRequestToQueryRequest is a ptr, if we keep passing pointer down the call then it would stay on
+	// heap until connection to websocket stays open
+	queryRequest := *queryRequestPtr
+	itr := q.tailQuery(r.Context(), &queryRequest)
+	stream := logproto.Stream{}
+
+	for itr.Next() {
+		stream.Entries = []logproto.Entry{itr.Entry()}
+		stream.Labels = itr.Labels()
+
+		err := conn.WriteJSON(stream)
+		if err != nil {
+			level.Error(util.Logger).Log("Error writing to websocket", fmt.Sprintf("%v", err))
+			helpers.LogError("writing close message to websocket", func() error {
```

Comment:
Can me make it the normal handling process instead of using `helpers.LogError`? It is useful when using defer, but I don't think it adds any benefits here. It only makes it complex.